io.c

```diff
@@ -2244,6 +2237,184 @@ static int io_sqe_files_register(struct io_ring_ctx *ctx, void __user *arg,
+		pr_err("%s: mapped max len is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len < arg.len) {
+		pr_err("%s: max len less than len", __func__);
+		goto err;
+	}
+
+	for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+		imu = &ctx->user_bufs[buf_index];
+		if (imu->ubuf <= (uintptr_t) arg.base &&
+		    (uintptr_t) arg.base + arg.len <= imu->ubuf + imu->len) {
+			offset = ((uintptr_t)arg.base - imu->ubuf) >> PAGE_SHIFT;
+			pr_info("found existing mapping %d, off %ld", buf_index, offset);
+			break;
+		}
+		if (imu->ubuf < (uintptr_t) arg.base + arg.len &&
+			(uintptr_t) arg.base < imu->ubuf + imu->len) {
+			pr_info("%s: intersects with existing mapping", __func__);
+			ret = -EEXIST;
+			goto err;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		existing = false;
+		for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+			pr_info("new mapping %d, ", buf_index);
+			if (ctx->user_bufs[buf_index].ubuf == 0)
+				break;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		pr_err("%s: all busy", __func__);
+		return -EBUSY;
+	}
+
+	imu = &ctx->user_bufs[buf_index];
+
+	/*
+         * Don't impose further limits on the size and buffer
+         * constraints here, we'll -EINVAL later when IO is
+         * submitted if they are wrong.
+         */
+	ret = -EFAULT;
+	if (!arg.base || !arg.len)
+		goto err;
+
+	/* arbitrary limit, but we need something */
+	if (arg.len > SZ_4G)
+		goto err;
+
+	ubuf = (unsigned long) arg.base;
+	end = (ubuf + arg.len) >> PAGE_SHIFT;
+	start = ubuf >> PAGE_SHIFT;
+	nr_pages = end - start;
+	max_nr_pages = arg.max_len >> PAGE_SHIFT;
+
+	if (existing) {
+		for (j = 0; j < nr_pages; ++j) {
+			if (imu->bvec[j + offset].bv_page) {
+				pr_err("%s: trying to overwrite existing mapping", __func__);
+				ret = -EEXIST;
+				goto err;
+			}
+		}
+	}
+
+	ret = -ENOMEM;
+	pages = kvmalloc_array(nr_pages, sizeof(struct page *),
+		GFP_KERNEL);
+	vmas = kvmalloc_array(nr_pages, sizeof(struct vm_area_struct *),
+		GFP_KERNEL);
+	if (!pages || !vmas) {
+		goto err;
+	}
+
+	if (!existing) {
+		imu->bvec = kvmalloc_array(max_nr_pages, sizeof(struct bio_vec),
+			GFP_KERNEL | __GFP_ZERO);
+		if (!imu->bvec) {
+			goto err;
+		}
+	}
+
+	ret = 0;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)
+	down_read(&current->mm->mmap_sem);
+#else
+	down_read(&current->mm->mmap_lock);
+#endif
+	pret = get_user_pages(ubuf, nr_pages, FOLL_WRITE, pages, vmas);
+	pr_info("%s: %ld bytes %d pages", __func__, arg.len, nr_pages);
+	if (pret < 0) {
+		ret = pret;
+	} else if (pret != nr_pages) {
+		ret = -EFAULT;
+	}
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)
+	up_read(&current->mm->mmap_sem);
+#else
+	up_read(&current->mm->mmap_lock);
+#endif
+	if (ret) {
+		/*
+                 * if we did partial map release any pages we did get
+                 */
+		if (pret > 0) {
+			for (j = 0; j < pret; j++)
+				put_page(pages[j]);
+		}
+		kvfree(imu->bvec);
+		imu->bvec = NULL;
```

Comment:
The above logic is incorrect. If there is any partial map failure, then in this attempt there was a part map failure. But the bvec may still hold pinned references to other successful mapping.

Reply:
added check for existing